src/github.com/matrix-org/dendrite/appservice/types/types.go

```diff
@@ -12,7 +12,46 @@
 
 package types
 
+import (
+	"sync"
+
+	"github.com/matrix-org/dendrite/common/config"
+)
+
 const (
 	// AppServiceDeviceID is the AS dummy device ID
 	AppServiceDeviceID = "AS_Device"
 )
+
+// ApplicationServiceWorkerState is a type that couples an application service,
+// a lockable condition as well as some other state variables, allowing the
+// roomserver to notify appservice workers when there are events ready to send
+// externally to application services.
+type ApplicationServiceWorkerState struct {
+	AppService config.ApplicationService
+	Cond       *sync.Cond
+	// Events ready to be sent
+	EventsReady *int
```

Comment:
I would really just have this as a `bool` that gets set when there are more events, and unset when `WaitForEvents` returns, as if there is an off by one error (or whatever) anywhere things will get weird quickly. If you use a boolean then it will self correct.